Sources/NIOResumableUpload/HTTPResumableUpload.swift

```diff
@@ -0,0 +1,401 @@
+                if let uploadSize = self.uploadSize {
+                    if let contentLength {
+                        conflict = complete && uploadSize != offset + contentLength
+                    } else {
+                        conflict = true
+                    }
+                } else {
+                    if let contentLength, complete {
+                        self.uploadSize = offset + contentLength
+                    }
+                    conflict = false
+                }
+            } else {
+                conflict = true
+            }
+            guard !conflict else {
+                self.detachUploadHandler(close: true)
+                self.destroyChannel(error: HTTPResumableUploadError.badResumption)
+                let response = HTTPResumableUploadProtocol.conflictResponse(
+                    offset: self.offset,
+                    complete: self.uploadComplete
+                )
+                self.respondAndDetach(response, handler: otherHandler)
+                return
+            }
+            self.requestIsCreation = false
+            self.requestIsComplete = complete
+            self.attachUploadHandler(otherHandler, channel: channel)
+        }
+    }
+
+    private func uploadCancellation() {
+        self.runInEventLoop {
+            self.detachUploadHandler(close: true)
+            self.destroyChannel(error: HTTPResumableUploadError.uploadCancelled)
+        }
+    }
+
+    private func receiveHead(handler: HTTPResumableUploadHandler, channel: Channel, request: HTTPRequest) {
+        self.eventLoop.preconditionInEventLoop()
+
+        switch HTTPResumableUploadProtocol.identifyRequest(request, in: self.context) {
+        case .notSupported:
+            let channel = self.createChannel(handler: handler, parent: channel)
+            channel.receive(.head(request))
+        case .uploadCreation(let complete, let contentLength):
+            self.requestIsCreation = true
+            self.requestIsComplete = complete
+            self.uploadSize = complete ? contentLength : nil
+            let resumePath = self.context.startUpload(self)
+            self.resumePath = resumePath
+
+            let informationalResponse = HTTPResumableUploadProtocol.featureDetectionResponse(resumePath: resumePath, in: self.context)
+            handler.writeAndFlush(.head(informationalResponse), promise: nil)
+
+            let strippedRequest = HTTPResumableUploadProtocol.stripRequest(request)
+            let channel = self.createChannel(handler: handler, parent: channel)
+            channel.receive(.head(strippedRequest))
+        case .offsetRetrieving:
+            if let path = request.path, let upload = self.context.findUpload(path: path) {
+                self.uploadHandler = nil
+                self.uploadHandlerChannel.withLockedValue { $0 = nil }
+                upload.offsetRetrieving(otherHandler: handler)
```

Comment:
We immediately call `detachUploadHandler` in `offsetRetrieving` so I don't think we should set the handler or channel to `nil` here (otherwise `detachUploadHandler` does nothing).